lib/datetime/datetime.ex

```diff
@@ -447,13 +447,17 @@ defimpl Timex.Protocol, for: DateTime do
           err
 
         %DateTime{} = datetime when shift != 0 ->
-          DateTime.add(datetime, shift, :microsecond, Timex.Timezone.Database)
+          datetime
+          |> DateTime.add(shift, :microsecond, Timex.Timezone.Database)
+          |> retain_precision(datetime)
 
         %DateTime{} = datetime ->
           datetime
 
-        {{ty, _, _}, %DateTime{} = orig} when ty in [:gap, :ambiguous] and shift != 0 ->
-          DateTime.add(orig, shift, :microsecond, Timex.Timezone.Database)
+        {{ty, _, _}, %DateTime{} = original} when ty in [:gap, :ambiguous] and shift != 0 ->
+          original
+          |> DateTime.add(shift, :microsecond, Timex.Timezone.Database)
+          |> retain_precision(datetime)
 
         {{ty, _a, _b} = amb, _} when ty in [:gap, :ambiguous] ->
           amb
```

Comment:
I would get rid of the case statement below as it seems completely unnecessary, but that may be outside of the scope of this PR. :shrug: :sweat_smile: